tensorflow_datasets/core/statistics.py

```diff
@@ -0,0 +1,133 @@
+"""Dataset statisitics util.
+"""
+
+import json
+import os
+
+from absl import logging
+from google.protobuf import json_format
+import tensorflow.compat.v2 as tf
+
+from tensorflow_datasets.core import dataset_builder
+from tensorflow_datasets.core import lazy_imports_lib
+from tensorflow_datasets.core import naming
+from tensorflow_datasets.core import splits as splits_lib
+from tensorflow_datasets.core import utils
+
+
+class StatisticsInfo:
```

Comment:
No problem 🙂
Will work on something else in the meantime.

Reply:
As I started looking into this, it seems that the reason why the `SplitInfo` is complicated was because the way `DatasetInfo` was written.

Currently, we need to keep `info.splits` and `info._info_proto` in sync, so updating `info.splits` require to notify `info._info_proto`. This makes the code hard to follow.
It's my fault for implementing this like that.

We should update `DatasetInfo` to be a `dataclasses.dataclass(frozen=True)`, and rather than using the proto as ground truth, use the DatasetInfo attributes. So we can update `info.splits` without having to update the proto.

So 
```py
class DatasetInfo:
  ...

  @property
  def as_proto(self):
    return self._info_proto

  @property
  def name(self):
    return self.as_proto.name
```
would become:
```py
@dataclasses.dataclass(frozen=True)
class DatasetInfo:
  ...

  name: str

  @property
  def as_proto(self):
    return dataset_info_pb2.DatasetInfo(
        name=self.name,
    )
```

There wouldn't be `_info_proto `, and `as_proto` would only be called when saving in `write_to_directory`, instead of for every attributes

Some implementations details:
* With `(frozen=True)`, because info is mutable, we can use `object.__setattr__(self, 'name', value)` when we need to update dataset info, like inside `read_from_directory`.
* The current __init__ would be changed as post_init to cleanup the arguments:


```py
@dataclasses.dataclass(frozen=True)
class _DatasetInfo:
  
  def __post_init__(self):
    object.__setattr__(self, 'version', self.builder.version)
    if self.description:
      object.__setattr__(self, 'description', utils.dedent(self.description))
```

You can look into this if you want but it's possible that I missed something again which would make this more complicated than expected.